smartmeter_datacollector/smartmeter/cosem.py

```diff
@@ -19,19 +37,59 @@ class RegisterCosem:
 
 
 class CosemConfig:
-    def __init__(self, id_obis: str, clock_obis: str, register_obis: List[RegisterCosem]) -> None:
-
-        self._id_obis = id_obis
-        self._clock_obis = clock_obis
+    def __init__(self, fallback_id: str, register_obis: List[RegisterCosem]) -> None:
+        self._id: Optional[str] = None
+        self._fallback_id = fallback_id
         self._register_obis = {r.obis: r for r in register_obis}
 
-    @property
-    def id_obis(self) -> str:
-        return self._id_obis
+    def retrieve_id(self, dlms_objects: Dict[str, Any]) -> str:
+        if self._id:
+            return self._id
+
+        id_obis = self._find_id_obis(dlms_objects)
```

Comment:
Right now the logic only consideres the ID from the first OBIS code present from `DEFAULT_ID_OBIS`. What if it is there but invalid? Should we loop through all available OBIS codes for the ID and try to find an ID from any of them?

Reply:
I would go that far. The checks in the `cosem.retrieve_id` method should handle the following error case:
```
DEBUG:smartmeter:40 [3, bytearray(b'\x01\x01\x02\x08\x00\xff'), 2, 0] 1: [3, bytearray(b'\x01\x01\x02\x08\x00\xff'), 2, 0]
DEBUG:smartmeter:1 0.0.42.0.0.255 2: [3, bytearray(b'\x01\x01\x05\x08\x00\xff'), 2, 0]
DEBUG:smartmeter:1 0.0.96.1.1.255 2: [3, bytearray(b'\x01\x01\x06\x08\x00\xff'), 2, 0]
DEBUG:smartmeter:8 0.0.1.0.0.255 2: [3, bytearray(b'\x01\x01\x07\x08\x00\xff'), 2, 0]
DEBUG:smartmeter:3 1.0.1.7.0.255 2: [3, bytearray(b'\x01\x01\x08\x08\x00\xff'), 2, 0]
DEBUG:smartmeter:3 1.0.2.7.0.255 2: [3, bytearray(b'\x01\x00\r\x07\x00\xff'), 2, 0]
DEBUG:smartmeter:3 1.0.3.7.0.255 2: 00 08 19 09 00 FF
```
This happens sometimes. Gurux could not correctly parse the data.
Either there is a ID obis object or not. However sometimes an ID cannot be parsed, but then the data is corrupt anyway and there is no need to find other IDs.